### PR TITLE
Edge Detection Post Process

### PR DESCRIPTION
**Pull request by UN1TOne**
https://forum.babylonjs.com/t/outline-post-process/50488

I wrote this article with my different accout.

This PR adds a new EdgeDetectionPostProcess to BabylonJS.
This post-process applies an edge detection effect based on normals and depth buffers from the geometry buffer renderer, providing developers with a versatile way to highlight object outlines.

Key Features
Configurable edge color, width, and intensity for flexible stylization.
Supports different render modes:

General
Normal
Depth
Only Outline
Uses Geometry Buffer Renderer to access normals and depth textures for precise edge detection.

Why this is needed?
This post-process enhances the visual rendering capabilities of BabylonJS by allowing developers to add stylized edge detection, which is commonly used in cartoon rendering, technical visualizations, and architectural previews.

How to Use
Enable the geometry buffer renderer and apply the post-process:

// edge detection post process
const edgeDetectionPP = new BABYLON.EdgeDetectionPostProcess("", scene, 1, camera);
edgeDetectionPP.samples = 4;
// Defines the intensity of the detected edges. Higher values result in more pronounced edges.
edgeDetectionPP.edgeIntensity = .3;
// Defines the width of the detected edges. Higher values result in thicker edges.
edgeDetectionPP.edgeWidth = .2;
// Defines the color of the detected edges
edgeDetectionPP.edgeColor = new BABYLON.Color3(0, 0, 0);

https://github.com/user-attachments/assets/f988c4cf-aef2-4ff3-a3de-0bdcf804631d

### CODE REVIEW

**Comment by UN1TOne** [comment]
I accidentally modified the package.json file, so I'm submitting the PR again 😥

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by noname0310** [comment]
It would be great if you could share your PG. This will make it easy to create tests.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15721/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15721/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15721/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15721/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15721/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15721/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15721/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15721/merge#BCU1XR#0

**Comment by UN1TOne** [comment]
https://playground.babylonjs.com/#HBSUDW


Since it hasn't been merged yet, an error might occur, so I'm providing the code separately.
Here's the test code

`var createScene = function () {
    var scene = new BABYLON.Scene(engine);
    scene.clearColor = BABYLON.Color4.FromHexString("#B8B897FF");

    var camera = new BABYLON.ArcRotateCamera("camera", BABYLON.Tools.ToRadians(90), BABYLON.Tools.ToRadians(65), 10, BABYLON.Vector3.Zero(), scene);
    camera.setTarget(new BABYLON.Vector3(0.07, 1.14, 0.06));
    camera.alpha = 1.8;
    camera.beta = 1.6;
    camera.radius = 4.8;
    camera.attachControl(canvas, true);
    camera.inertia = 0.4;
    camera.wheelDeltaPercentage = 0.3;
    camera.panningSensibility = 10000;
    camera.minZ = 0.01;
    camera.maxZ = 1000;
    camera.zoomToMouseLocation = true;

    var light = new BABYLON.HemisphericLight("light", new BABYLON.Vector3(0, 1, 0), scene);
    light.intensity = 0.7;

    // edge detection post process
    const edgeDetectionPP = new BABYLON.EdgeDetectionPostProcess("", scene, 1, camera);
    edgeDetectionPP.samples = 4;
    // Defines the intensity of the detected edges. Higher values result in more pronounced edges.
    edgeDetectionPP.edgeIntensity = .3;
    // Defines the width of the detected edges. Higher values result in thicker edges.
    edgeDetectionPP.edgeWidth = .2;
    // Defines the color of the detected edges
    edgeDetectionPP.edgeColor = new BABYLON.Color3(0, 0, 0);

    // This can help anti-aliasing but it can cause performance issues
    engine.setHardwareScalingLevel(.7);

    var defaultPipeline = new BABYLON.DefaultRenderingPipeline("default", true, scene, [camera]);
    defaultPipeline.fxaaEnabled = true;
    defaultPipeline.samples = 4;

    BABYLON.SceneLoader.ImportMesh("", "scenes/BrainStem/", "BrainStem.gltf", scene);

    return scene;
};`

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15721/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15721/merge/testResults/webgl2playwright/index.html

**Comment by noname0310** [comment]
> Since it hasn't been merged yet, an error might occur, so I'm providing the code separately.

You can see it with the snapshot link
https://playground.babylonjs.com/?snapshot=refs/pull/15721/merge#HBSUDW

The postProcesses package doesn't currently have a tests folder, but it would be nice to have tests to ensure backward compatibility of the many postprocesses that will be added in the future. We could discuss it in the forums, but it seemed to be needed for this PR, so I'm mentioning it here, sorry.@RaananW

**Comment by RaananW** [comment]
> > Since it hasn't been merged yet, an error might occur, so I'm providing the code separately.
> 
> You can see it with the snapshot link https://playground.babylonjs.com/?snapshot=refs/pull/15721/merge#HBSUDW
> 
> The postProcesses package doesn't currently have a tests folder, but it would be nice to have tests to ensure backward compatibility of the many postprocesses that will be added in the future. We could discuss it in the forums, but it seemed to be needed for this PR, so I'm mentioning it here, sorry.@RaananW

Visualization tests are global, so you can add a new vis-test that will use the new post process. If you want unit tests you can add new ones as well, but I assume you refer to vis-tests

**Comment by Popov72** [comment]
(recreating my comment from the deleted PR)

I think it would be better suited to our [PostProcess package](https://doc.babylonjs.com/setup/frameworkPackages/CDN). What do you think @sebavan / @deltakosh ?

**Comment by deltakosh** [comment]
I tend to agree with @Popov72 

@UN1TOne would you be so kind to move it here: https://github.com/BabylonJS/Babylon.js/tree/master/packages/dev/postProcesses/src

**Comment by UN1TOne** [comment]
I've changed the path of the code I worked on as you specified (https://github.com/BabylonJS/Babylon.js/tree/master/packages/dev/postProcesses/src)

@Popov72 @deltakosh

File: packages/dev/core/src/ShadersWGSL/edgeDetection.fragment.fx [hunk 1]
```diff
@@ -0,0 +1,69 @@
+precision highp float;
```
Comment (noname0310) [inline packages/dev/core/src/ShadersWGSL/edgeDetection.fragment.fx:1]: This file should also be moved

Comment (UN1TOne) [reply packages/dev/core/src/ShadersWGSL/edgeDetection.fragment.fx]: I should've checked again
Sorry. It's my mistake😅

Comment (UN1TOne) [reply packages/dev/core/src/ShadersWGSL/edgeDetection.fragment.fx]: But it seems that the pipeline times out after changing the path.
What am I have to do?

**Review by sebavan — APPROVED** [review]

**Pull request merged**